[IMP] website_event: use turnstile when available

reth-odoo opened the pull request:
We recently enabled recaptcha on the registration form in [1]

Turnstile can be implemented on top of it by just invoking the turnstile script in js.

As it's implicitly checked when recaptcha is checked, this also avoids any issue with the form potentially not working if turnstile is enabled.

[1] bafa915f85fb8fc6ca2ae7d194a4593cb2463c2e

task-4335141

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/192827.png)](https://mergebot.odoo.com/odoo/odoo/pull/192827)

thjo-odoo commented on addons/website_event/static/src/js/website_event.js:
```diff
@@ -100,6 +103,40 @@ var EventRegistrationForm = publicWidget.Widget.extend({
+            return false;
+        }
+        this._removeTurnstile();
+        const div = document.createElement("div");
+        div.classList.add("s_turnstile");
+        div.classList.add("cf-turnstile");
+        div.classList.add("float-end")
+        div.setAttribute("data-sitekey", session.turnstile_site_key);
+        div.setAttribute("data-appearance", "interaction-only");
+        div.setAttribute("data-appearance", "always");
+        div.setAttribute("data-response-field-name", "turnstile_captcha");
+        div.setAttribute("data-action", "recaptcha_token_response");
+        div.setAttribute("data-error-callback", "(code) => throw new Error('Turnstile error: '+ code)");
```
All the more reason to have a template for this.

reth-odoo replied:
Yep, added a parametrized template in website_cf_turnstile directly

thjo-odoo commented on addons/website_event/static/src/js/website_event.js:
```diff
@@ -100,6 +103,40 @@ var EventRegistrationForm = publicWidget.Widget.extend({
+        div.setAttribute("data-error-callback", "(code) => throw new Error('Turnstile error: '+ code)");
+
+        const modalFooter = form.querySelector("div.modal-footer");
+        modalFooter.insertBefore(div, modalFooter.firstChild);
```
This should be the same
```suggestion
        modalFooter.prepend(div);
```

thjo-odoo reviewed (commented):
Hello @reth-odoo,
Here's the COM part, and it's the same as for the ENT PR. The same gripe and all.

The only thing is that I saw your commit message and I think it's a bit long for some lines. 
Cheers!

thjo-odoo commented on addons/website_cf_turnstile/static/src/js/turnstile.js:
```diff
@@ -13,14 +14,15 @@ publicWidget.registry.s_website_form.include({
+                    sitekey: session.turnstile_site_key,
+                });
+                const turnstileScript = renderToElement("website_cf_turnstile.turnstile_remote_script");
+                const turnstileErrorGlobalScript = $(`<script class="s_turnstile">
```
JQuery is still authorised on the frontend? This is more as to avoid adding some more when we could use something else.

reth-odoo replied:
@thjo-odoo migrated the script to javascript only while still using a script element, just in case.

Feature seems to still work fine.

thjo-odoo commented on addons/website_cf_turnstile/static/src/js/turnstile.js:
```diff
@@ -29,8 +31,9 @@ publicWidget.registry.s_website_form.include({
-                    <script class="s_turnstile" src="https://challenges.cloudflare.com/turnstile/v0/api.js"></script>
-                `).insertAfter('.s_website_form_send, .o_website_form_send');
+                `);
+                $(turnstileContainer).add(turnstileErrorGlobalScript).add($(turnstileScript))
```
`turnstileContainer` is, I think, already an element so you could add them like you would in regular JS.

reth-odoo replied:
I tried to keep it logically close to the original in case some jquery sideeffect had a visible impact. But indeed we can convert to standard js dom api and see if framework has an objection later.

thjo-odoo reviewed (commented):
@reth-odoo 
Here's some other thoughts on this.

thjo-odoo commented on addons/website_cf_turnstile/static/src/js/turnstile.js:
```diff
@@ -12,25 +13,32 @@ publicWidget.registry.s_website_form.include({
+                const turnstileScript = renderToElement("website_cf_turnstile.turnstile_remote_script");
+
+                const sendButton = document.querySelector(".s_website_form_send, .o_website_form_send");
+                sendButton.parentNode.insertBefore(turnstileContainer, sendButton.nextSibling);
+                sendButton.parentNode.insertBefore(turnstileErrorGlobalScript, sendButton.nextSibling);
+                sendButton.parentNode.insertBefore(turnstileScript, sendButton.nextSibling);
```
Aren't all these calls simply this:
```suggestion
                sendButton.after(turnstileContainer, turnstileErrorGlobalScript, turnstileScript);
```

reth-odoo replied:
:o nice, I thought it was weird for there to not be an insertAfter in the dom api :D

patApplied

thjo-odoo reviewed (commented):
@reth-odoo 
Last one and we should be good to go

reth-odoo commented:
Ok. 

-Reworked to load website_cf_turnstile dynamically if session key is set. (if website_cf_turnstile is not loaded, we'll just do nothing so it's fine).

- Reworked to call website_cf_turnstile methods for widget creation/initialization, and just manually edit style where needed.

-Added an explicit render method that works with the output of the standard `addTurnstile`, so that we can keep rendering new turnstiles even if the remote script is already loaded. (should be called manually for stable at least).

- Removed the limit of 1 turnstile, and added unique callbacks to properly support multiple turnstile checks per page.

reth-odoo commented on addons/website_cf_turnstile/static/src/js/turnstile.js:
```diff
@@ -2,44 +2,58 @@
-                </script>
-                <script class="s_turnstile" src="https://challenges.cloudflare.com/turnstile/v0/api.js"></script>
-            `);
+            const callbackUUID = `${action}-${Date.now()}-${Math.random()}`;
```
Responded on your PR for an example but the callback uuid is more so that existing callbacks defined here are called when their relevant widget succeeds, rather than any widget.
Not so much to be able to add a custom callback (which could be done by modifying the result of this method before inserting it into the dom/rendering it, anyway)

reth-odoo replied:
So I also went back to using full-fledged functions instead of closures so the widget can bind "this" properly. As I thought `this.wrapper` was some archane "the node of this script" magic originally ^^'

Works fine now

reth-odoo replied:
@bram1000 :wave: mind having a (hopefully) final look for the web part?

reth-odoo commented on addons/website_cf_turnstile/static/src/js/turnstile.js:
```diff
@@ -61,9 +74,50 @@ export const turnStile = {
         spinner.classList.add("fa", "fa-refresh", "fa-spin", "turnstile-spinner");
+        return spinner;
+    },
+
+    /**
+     * same as addSpinner but does not set innerText
+     */
+    addSpinnerNoMangle(button) {
+        const spinner = this._createSpinner();
+        spinner.classList.add("me-1");
+        button.disabled = true;
+        button.classList.add("disabled");
+        button.prepend(spinner);
+    },
```
Fixed
This thing breaks on runbot but not locally so it was fun :smiling_face_with_tear:
Probably an override of event adds line breaks that are normally invisible but innerText belows transforms them into <br> which are very visible
